go-controller/pkg/ovn/egressgw.go

```diff
@@ -197,99 +320,101 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod, namespace string) {
-func (oc *Controller) deleteGWRoutesForNamespace(nsInfo *namespaceInfo) {
-	if nsInfo == nil {
-		return
-	}
-
+func (oc *Controller) deleteGWRoutesForNamespace(namespace string) {
 	// TODO(trozet): batch all of these with ebay bindings
-	for podIP, gwToGr := range nsInfo.podExternalRoutes {
-		for gw, gr := range gwToGr {
-			if utilnet.IsIPv6String(gw) != utilnet.IsIPv6String(podIP) {
-				continue
-			}
-			mask := GetIPFullMask(podIP)
-			node := util.GetWorkerFromGatewayRouter(gr)
-			if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), node, false); err != nil {
-				klog.Error(err)
-			}
-			_, stderr, err := util.RunOVNNbctl("--if-exists", "--policy=src-ip",
-				"lr-route-del", gr, podIP+mask, gw)
-			if err != nil {
-				klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
-			} else {
-				delete(nsInfo.podExternalRoutes, podIP)
-			}
+	routeInfos := oc.getRouteInfosForNamespace(namespace)
+	for _, routeInfo := range routeInfos {
+		for podIP, gwToGr := range routeInfo.routes {
+			for gw, gr := range gwToGr {
+				if utilnet.IsIPv6String(gw) != utilnet.IsIPv6String(podIP) {
+					continue
+				}
+				mask := GetIPFullMask(podIP)
+				node := util.GetWorkerFromGatewayRouter(gr)
+				if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), node, false); err != nil {
+					klog.Error(err)
+				}
+				_, stderr, err := util.RunOVNNbctl("--if-exists", "--policy=src-ip",
+					"lr-route-del", gr, podIP+mask, gw)
+				if err != nil {
+					klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+				} else {
+					delete(routeInfo.routes, podIP)
+				}
 
-			portPrefix, err := oc.extSwitchPrefix(node)
-			if err != nil {
-				klog.Infof("Failed to find ext switch prefix for %s %v", node, err)
-				continue
+				portPrefix, err := oc.extSwitchPrefix(node)
+				if err != nil {
+					klog.Infof("Failed to find ext switch prefix for %s %v", node, err)
+					continue
+				}
+				cleanUpBFDEntry(gw, gr, portPrefix)
 			}
-			cleanUpBFDEntry(gw, gr, portPrefix)
 		}
+		routeInfo.Unlock()
 	}
-	nsInfo.routingExternalGWs = gatewayInfo{}
 }
 
 // deleteGwRoutesForPod handles deleting all routes to gateways for a pod IP on a specific GR
-func (oc *Controller) deleteGWRoutesForPod(namespace string, podIPNets []*net.IPNet) {
-	// delete src-ip cached route to GR
-	nsInfo, nsUnlock := oc.getNamespaceLocked(namespace, false)
-	if nsInfo == nil {
+func (oc *Controller) deleteGWRoutesForPod(name ktypes.NamespacedName, podIPNets []*net.IPNet) {
+	routeInfo := oc.deleteRouteInfoLocked(name)
+	if routeInfo == nil {
 		return
 	}
-	defer nsUnlock()
+	defer routeInfo.Unlock()
 
 	for _, podIPNet := range podIPNets {
 		pod := podIPNet.IP.String()
-		if gwToGr, ok := nsInfo.podExternalRoutes[pod]; ok {
+		if gwToGr, ok := routeInfo.routes[pod]; ok {
 			if len(gwToGr) == 0 {
-				delete(nsInfo.podExternalRoutes, pod)
+				delete(routeInfo.routes, pod)
```

Comment:
hmm when we trigger `oc.deleteRouteInfoLocked(name)` if all things go well, we will be deleting the whole `*externalRouteInfo` from `egressGWCache` correct? So `routeInfo` will be nil always? what's the point of deleting this now from rotueInfo.routes? I'd expect that whole routes map to be gone from the cache for this pod? Feels to me that the key from egressGWCache is already gone and we are playing with its value here.

Reply:
routeInfo is returned (not nil), but you are right its not longer in the cache. Deleting the routes from the routeInfo in this case doesn't really do anything. However, if we look at it from an API perspective, the code might change in the future, and i think being more consistent here has almost 0 cost so we might as well clean up the routes.

Reply:
turns out we actually need this (although it was wrong initially) for the second commit